trillian/integration/copier.go

```diff
@@ -0,0 +1,211 @@
+		if err := targetPool.AppendCertsFromPEMFile(pemFile); err != nil {
+			return nil, fmt.Errorf("failed to read trusted roots for target log: %v", err)
+		}
+	}
+	if glog.V(4) {
+		for _, cert := range targetPool.RawCertificates() {
+			glog.Infof("target root cert: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+		}
+	}
+
+	seenOverlap := false
+	srcRoots, err := client.GetAcceptedRoots(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read trusted roots for source log: %v", err)
+	}
+	sourcePool := ctfe.NewPEMCertPool()
+	for _, root := range srcRoots {
+		cert, err := x509.ParseCertificate(root.Data)
+		if x509.IsFatal(err) {
+			glog.Warningf("Failed to parse root certificate from source log: %v", err)
+			continue
+		}
+		glog.V(4).Infof("source log root cert: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+		sourcePool.AddCert(cert)
+		if targetPool.Included(cert) {
+			glog.V(3).Infof("source log root cert is accepted by target: %x Subject: %v", sha256.Sum256(cert.Raw), cert.Subject)
+			seenOverlap = true
+		}
+	}
+	if !seenOverlap {
+		return nil, fmt.Errorf("failed to find any overlap in accepted roots for target %s", cfg.Prefix)
+	}
+
+	if startIndex == -1 {
+		// Pick a random start point in the source log's tree.
+		sth, err := client.GetSTH(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get STH for source log: %v", err)
+		}
+		startIndex = rand.Intn(int(sth.TreeSize))
+		glog.Infof("starting CopyChainGenerator from index %d (of %d) in source log", startIndex, sth.TreeSize)
+	}
+
+	generator := &CopyChainGenerator{
+		start:       start,
+		limit:       limit,
+		targetRoots: targetPool,
+		sourceRoots: sourcePool,
+		certs:       make(chan []ct.ASN1Cert, bufSize),
+		precerts:    make(chan []ct.ASN1Cert, bufSize),
+	}
+
+	// Start two goroutines to scan the source log for certs and precerts respectively.
+	fetchOpts := scanner.FetcherOptions{
+		BatchSize:     bufSize,
+		ParallelFetch: 1,
+		Continuous:    true,
+		StartIndex:    int64(startIndex),
+	}
+	certFetcher := scanner.NewFetcher(client, &fetchOpts)
+	go certFetcher.Run(ctx, func(batch scanner.EntryBatch) {
+		generator.processBatch(batch, generator.certs, ct.X509LogEntryType)
+	})
+
+	precertFetcher := scanner.NewFetcher(client, &fetchOpts)
+	go precertFetcher.Run(ctx, func(batch scanner.EntryBatch) {
+		generator.processBatch(batch, generator.precerts, ct.PrecertLogEntryType)
+	})
+
+	return generator, nil
+}
+
+// processBatch extracts chains of the desired type from a batch of entries and sends
+// them down the channel.  May block on the channel consumer.
+func (g *CopyChainGenerator) processBatch(batch scanner.EntryBatch, chains chan []ct.ASN1Cert, eType ct.LogEntryType) {
+	glog.V(2).Infof("processBatch(%d): examine batch [%d, %d)", eType, batch.Start, int(batch.Start)+len(batch.Entries))
+	for i, entry := range batch.Entries {
+		index := batch.Start + int64(i)
+		entry, err := ct.RawLogEntryFromLeaf(index, &entry)
```

Comment:
Looks like it might ignore err?

Reply:
Shifted to a more normal check of `err != nil`  (exactly one of `entry` and `err` is non-`nil`)